lib/pry/input_completer.rb

```diff
@@ -239,4 +227,39 @@ def build_path(input)
     end
     return path, input
   end
+
+  private
+
+  def self.fetch_obj_space_collection
```

Comment:
@kvokka wdyt on separating mutex logic, checking for stale cache and preparing methods list? I hope it'll be easier to read and understand.

```ruby
def all_available_methods_cached
  @mutext.sync do
    @all_available_methods_cached = all_available_methods if stale_methods_cache?
    @all_available_methods_cached
  end
end

def stale_methods_cache?
  # check that modules count changed
end

def all_available_methods
  # Fetch from ObjectSpace
end
```